securedrop/tests/functional/functional_test.py

```diff
@@ -33,19 +34,50 @@ class FunctionalTest():
 
     def _unused_port(self):
         s = socket.socket()
-        s.bind(("localhost", 0))
+        s.bind(('127.0.0.1', 0))
         port = s.getsockname()[1]
         s.close()
         return port
 
+
+    def _create_tor_browser_webdriver(self, abs_log_file_path):
+        # Read the TorBrowser location from Ansible
+        with open('/opt/.tbb_path_file') as f:
+           path_to_tbb = f.readline().strip()
+        path_to_tbb = path_to_tbb + os.path.sep + "tor-browser_en-US"
+        path_to_tbb = os.path.abspath(path_to_tbb)
+
+        # Don't use Tor when reading from localhost, and turn off private browsing.
+        # We need to turn off private browsing because we won't be able to access the browser's cookies in private browsing mode.
+        # Since we use session cookies in SD anyway (in private browsing mode all cookies are set as session cookies),
+        # this should not affect session lifetime.
+        pref_dict = {
+                     'network.proxy.no_proxies_on': '127.0.0.1',
+                     'browser.privatebrowsing.autostart': False
+                    }
+        driver = TorBrowserDriver(path_to_tbb,
+                                  pref_dict=pref_dict,
+                                  tbb_logfile_path=abs_log_file_path)
+        return driver
+
+    def _create_firefox_webdriver(self, abs_log_file_path):
+        with open(abs_log_file_path, 'a') as f:
+            firefox = firefox_binary.FirefoxBinary(log_file=f)
```

Comment:
The docs you linked say to send a file object, so what you're doing here is actually correct--my bad. I just got confused because `TorBrowserDriver` takes a string for the `log_file` keyword argument, whereas `FirefoxBinary` takes a file object--the former would ideally be made consistent with the latter since the latter is its upstream, but irrelevant to what we want to do here, and no big deal.

I think you'll need to back out of using `with` here, since our `FirefoxBinary` initializer is going to create an attribute in our `firefox` instance that is an obj ref to the same file object that `fh` is an obj ref to. When the `with` context manager exits, it will call `fh.close()`, which might prevent logging later on... although it seems like that would cause some fatal exceptions somewhere. I haven't looked at the code in a while, but can do that when I'm done with the current thing. Anyway, here's what I mean:

```
In [25]: with open('README.MD') as fh:
    ...:     x = fh
    ...:

In [26]: x.write('yo')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-26-f0d743a38770> in <module>()
----> 1 x.write('yo')

ValueError: I/O operation on closed file.
```